### PR TITLE
[processor/transform] Add functionality to break up a summary metric into Sum metrics.

### LINKED ISSUE #10824: [receiver/prometheus] Convert Summary metric into two monotonic sums.
**Is your feature request related to a problem? Please describe.**
When collecting Vault metrics using prometheus, a `summary` metric type is collected for `vault_gcs_list`.  

Example:
```json
{
  "name":"vault_gcs_list",
  "description":"vault_gcs_list",
  "summary":{
     "dataPoints":[
        {
           "startTimeUnixNano":"1654606313865000000",
           "timeUnixNano":"1654606313865000000",
           "count":"95",
           "sum":4811.074604034424,
           "quantileValues":[
              {
                 "quantile":0.5,
                 "value":"NaN"
              },
              {
                 "quantile":0.9,
                 "value":"NaN"
              },
              {
                 "quantile":0.99,
                 "value":"NaN"
              }
           ]
        }
     ]
  }
}
```
I would like to split the `sum` and the `count`  into two separate monotonic sums.

**Describe the solution you'd like**
Perhaps an enhancement to the `transformprocessor` to allow a change from summary to other metric types.

### PR DESCRIPTION
**Pull request by Mrod1598**
**Description:** <Describe what has changed.>
Resolves #10824

**Testing:** <Describe what testing was performed and which tests were added.>
Added tests for new functions. Tested processor against a real vault instance as well using the prometheus receiver.

**Documentation:** <Describe the documentation added.>
Added To README describing the new functions.

### CODE REVIEW

**Comment by TylerHelmuth** [comment]
@Mrod1598 I haven't reviewed this yet but I noticed the new functions are in `functions.go`. As part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10101 all new functions should be in their own files.  Checkout https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/11019 if you want to see the future CONTRIBUTING guide for functions.

File: processor/transformprocessor/README.md
Comment (antonblock) [inline processor/transformprocessor/README.md:48]:
```suggestion
- `convert_summary_count_val_to_sum(aggregation_temporality, is_monotonic)` - Creates a new Sum metric out of incoming metrics of type "Summary" with a "Count" Value. Noop for metrics that are not of type "Summary".
```
Worth adding to be consistent with the other convert functions

Comment (Mrod1598) [reply processor/transformprocessor/README.md]: Updated

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:47]: How'd you land at this name?  Whats the normal procedure for naming metrics created in a processor?  Should the function take a "name" parameter so that users can name the metric whatever they want?  Would be nice if it could be optional. Also applies to the other function.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:46]: Will the description translate correctly?  Should the function take a "description" parameter so that users can set the description to be whatever they want?  Would be nice if it could be optional. Also applies to the other function.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:58]: Do StartTimestamp and Timestamp need set on the new data point based on the summary data point?  Also applies to the other function.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go:60]:
```suggestion
func Test_ConvertSummarySumValToSum(t *testing.T) {
```

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go:67]: Need test scenarios for setting "cumulative", is_monotonic=true, and NOOP for other data types. Also applies to the other function test.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go:116]: Should have a `_validation` test that these the aggregation temporality switch statement failing. Also applies to the other function test.

File: processor/transformprocessor/internal/metrics/func_convert_summary_sum_val_to_sum_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_sum_val_to_sum_test.go:27]:
```suggestion
func Test_ConvertSummaryCountValToSum(t *testing.T) {
```

**Review by TylerHelmuth — COMMENTED** [review]
Thanks for tackling this.  Please add some integration tests to `internal/metrics/processor_test.go` for the 2 new functions.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:46]: probably wouldn't be completely correct but figured it was better than not transferring anything. Again, is there support for optional params?

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:47]: The grammar doesn't know how to do optional params yet 🙁. I guess the name could be changed in a later set function.

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:47]: The new name should be documented in the README.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:46]: Could probably be set with the set function later.  Please document all fields that are copied.

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:45]: I realized this modifies the metrics list that is being iterated though in processor.go.  AppendEmpty will add to the end of the list and the list's length is check on each iteration of the loop so I think the new metric would also run through each function right?  Can you please add some tests to the metric's processor_test.go that confirms this.

Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:45]: It does, added a test and a note in the doc

File: processor/transformprocessor/internal/metrics/func_convert_summary_sum_val_to_sum_test.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_sum_val_to_sum_test.go]: Fixed

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go:232]: Added

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go]: Added

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum_test.go]: Fixed

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:59]: Updated to transfer those as well

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:46]: Documented

File: processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go
Comment (Mrod1598) [reply processor/transformprocessor/internal/metrics/func_convert_summary_count_val_to_sum.go:47]: Added

File: processor/transformprocessor/README.md
Comment (TylerHelmuth) [inline processor/transformprocessor/README.md:51]:
```suggestion
- `convert_summary_sum_val_to_sum(aggregation_temporality, is_monotonic)` - Creates a new Sum metric out of incoming metrics of type "Summary" with a "Sum" Value. Noop for metrics that are not of type "Summary". The name for the new metric with be `<summary metric>_sum`. The fields that are copied are: `timestamp`, `starttimestamp`, `attibutes`, and `description`.  The new metric that is created will be passed to all functions in the metrics queries list.  Function conditions will apply.
```

The processor applies every function (that passes the condition check) to every data point.  So the metric created by this function will then be passed to every function defined in the queries list.  Also please add the statement to the other function.

**Review by TylerHelmuth — COMMENTED** [review]
nit

File: processor/transformprocessor/README.md
Comment (Mrod1598) [reply processor/transformprocessor/README.md]: Updated

**Review by TylerHelmuth — APPROVED** [review]
Thanks!

**Comment by Mrod1598** [comment]
@TylerHelmuth thanks for Reviewing! Can this merge or are we waiting on another review?

**Comment by TylerHelmuth** [comment]
We like to get at least 2 approvals for most things. @anuraaga @open-telemetry/collector-contrib-approvers please review.

**Review by djaglowski — APPROVED** [review]
LGTM

**Pull request merged**
